mypyc/irbuild/format_str_tokenizer.py

```diff
@@ -186,3 +191,51 @@ def join_formatted_strings(builder: IRBuilder, literals: Optional[List[str]],
 
     result_list[0] = Integer(len(result_list) - 1, c_pyssize_t_rprimitive)
     return builder.call_c(str_build_op, result_list, line)
+
+
+def convert_expr_to_bytes(builder: IRBuilder, format_ops: List[FormatOp],
+                          exprs: List[Expression], line: int) -> Optional[List[Value]]:
+    """Convert expressions into bytes literals with the guidance
+    of FormatOps."""
+    if len(format_ops) != len(exprs):
+        return None
+
+    converted = []
+    for x, format_op in zip(exprs, format_ops):
+        node_type = builder.node_type(x)
+        # conversion type 's' is an alias of 'b' in bytes formatting
+        if format_op == FormatOp.BYTES or format_op == FormatOp.STR:
+            if is_bytes_rprimitive(node_type):
+                var_bytes = builder.accept(x)
+            else:
+                return None
+        else:
+            return None
+        converted.append(var_bytes)
+    return converted
+
+
+def join_formatted_bytes(builder: IRBuilder, literals: List[str],
+                         substitutions: List[Value], line: int) -> Value:
+    """Merge the list of literals and the list of substitutions
+    alternatively using 'bytes_build_op'.
+
+    The literals should only contains ascii literal objects.
```

Comment:
Do these need to be ascii? Is it sufficient for them to be arbitrary bytes literals?